rocAL/rocAL/source/pipeline/tensor.cpp

```diff
@@ -0,0 +1,374 @@
+    return 0;
+}
+
+int rocALTensor::create_from_handle(vx_context context)
+{
+    if (_vx_handle)
+    {
+        WRN("Tensor object create method is already called ")
+        return -1;
+    }
+
+    _context = context;
+    vx_enum tensor_data_type = interpret_tensor_data_type(_info.data_type());
+    unsigned num_of_dims = _info.num_of_dims();
+    vx_size stride[num_of_dims];
+    void *ptr[1] = {nullptr};
+
+    stride[0] = tensor_data_size(_info.data_type());
+    for(unsigned i = 1; i < num_of_dims; i++)
+        stride[i] = stride[i - 1] * _info.dims().at(i - 1);
+
+    vx_size dims[num_of_dims];
+    for(unsigned i = 0; i < num_of_dims; i++)
+    {
+        dims[i] = _info.dims().at(i);
+    }
+    _vx_handle = vxCreateTensorFromHandle(_context, num_of_dims, dims, tensor_data_type, 0, stride, ptr, vx_mem_type(_info._mem_type));
+    vx_status status;
+    if ((status = vxGetStatus((vx_reference)_vx_handle)) != VX_SUCCESS)
+        THROW("Error: vxCreateTensorFromHandle(input: failed " + TOSTR(status))
+    _info._type = rocALTensorInfo::Type::HANDLE;
+    // shobi to be checked
+    // if(_info._data_size == 0)
+    //     _info._data_size = stride[3] * _info.channels(); // since data size is set while initializing info
+    return 0;
+}
+
+int rocALTensor::create(vx_context context)
+{
+    if (_vx_handle)
+        return -1;
+
+    _context = context;
+
+    unsigned num_of_dims = _info.num_of_dims();
+    vx_size dims[num_of_dims];
+    for(unsigned i = 0; i < num_of_dims; i++)
+    {
+        dims[i] = _info.dims().at(i);
+    }
+    vx_status status;
+    vx_enum tensor_data_type = interpret_tensor_data_type(_info.data_type());
+    _vx_handle = vxCreateTensor(context, num_of_dims, dims, tensor_data_type, 0);
+    if ((status = vxGetStatus((vx_reference)_vx_handle)) != VX_SUCCESS)
+        THROW("Error: vxCreateTensor(input: failed " + TOSTR(status))
+    _info._type = rocALTensorInfo::Type::REGULAR;
+    return 0;
+}
+#if ENABLE_OPENCL
+unsigned rocALTensor::copy_data(cl_command_queue queue, unsigned char *user_buffer, bool sync)
+{
+    if (_info._type != rocALTensorInfo::Type::HANDLE)
+        return 0;
+
+    if (_info._mem_type == RocalMemType::OCL)
+    {
+
+        cl_int status;
+        if ((status = clEnqueueReadBuffer(queue,
+                                          (cl_mem)_mem_handle,
+                                          sync ? (CL_TRUE) : CL_FALSE,
+                                          0,
+                                          _info.data_size(),
+                                          user_buffer,
+                                          0, nullptr, nullptr)) != CL_SUCCESS)
+            THROW("clEnqueueReadBuffer failed: " + TOSTR(status))
+    }
+    else
+    {
+        memcpy(user_buffer, _mem_handle, _info.data_size());
+    }
+    return _info.data_size();
+}
+unsigned rocALTensor::copy_data(cl_command_queue queue, cl_mem user_buffer, bool sync)
```

Comment:
Is this required?
We are returning 0.

Reply:
It is not required for our current tensor implementations, The image.cpp had copy_data API which returned the size